### PR TITLE
ISPN-11393 statistics, metrics, jmx

### PR DESCRIPTION
**Pull request by oraNod**
https://issues.redhat.com/browse/ISPN-11393 

Needs backport to 10.1.x

### CODE REVIEW

**Comment by anistor** [comment]
Looking...

File: documentation/src/main/asciidoc/topics/code_examples/ConfigStats.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+GlobalConfiguration globalConfig = new GlobalConfigurationBuilder()
+  .statistics().enable()
+  .metrics().gauges().histograms()
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/code_examples/ConfigStats.java]: .metrics().gauges(true).histograms(true)

File: documentation/src/main/asciidoc/stories/assembly_server_statistics.adoc [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[id='statistics']
+:context: server
+= Configuring Statistics, Metrics, and JMX
+Enable statistics that {brandname} exports to a metrics endpoint or via JMX
```
Comment (anistor) [inline documentation/src/main/asciidoc/stories/assembly_server_statistics.adoc]: ``` exports to a metrics endpoint```  --> ``` exports to a MicroProfile Metrics endpoint```

File: documentation/src/main/asciidoc/topics/code_examples/HotRodClientInstance.java [hunk 3]
```diff
@@ -3,7 +3,6 @@
 cb.marshaller(new org.infinispan.commons.marshall.ProtoStreamMarshaller())
   .statistics()
       .enable()
-      .jmxDomain("org.infinispan")
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/code_examples/HotRodClientInstance.java]: jmxDomain is not deprecated in HR client. This line can be kept, but maybe I would use something different than "org.infinispan", which is the default anyway, so it's not useful.

File: core/src/main/resources/schema/infinispan-config-11.0.xsd [hunk 4]
```diff
@@ -270,12 +270,16 @@
...
+          <xs:documentation>
+            Configures MicroProfile metrics.
+          </xs:documentation>
         </xs:annotation>
       </xs:element>
       <xs:element name="jmx" type="tns:jmx" minOccurs="0">
         <xs:annotation>
-          <xs:documentation>Defines JMX management details.</xs:documentation>
+          <xs:documentation>
+            Enables and configures JMX monitoring.
```
Comment (anistor) [inline core/src/main/resources/schema/infinispan-config-11.0.xsd]: monitoring and management

File: documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+[source,java]
+----
+GlobalConfiguration globalConfig = new GlobalConfigurationBuilder()
+  .statistics().enable() <1>
+  .metrics().gauges().histograms() <2>
+  .build();
+----
+
+<1> exports statistics about the Cache Manager.
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc]: ```exports statistics about ...``` --> ```computes statistics about ...```

File: documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+<2> generates gauge and histogram metrics.
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc]: ```generates gauge and histogram metrics```  --> ```exports the collected statistics as gauge and histogram metrics```

File: documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+</cache-container>
+----
+
+.Programmatically
+
+[source,java]
+----
+GlobalConfiguration globalConfig = new GlobalConfigurationBuilder()
+  .statistics().enable() <1>
+  .metrics().gauges().histograms() <2>
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc]: .metrics().gauges(true).histograms(true)

File: documentation/src/main/asciidoc/topics/proc_enabling_statistics.adoc [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  .build();
+
+ ...
+
+Configuration config = new ConfigurationBuilder()
+  .statistics().enable() <2>
+  .build();
+----
+
+<1> exports statistics about the Cache Manager.
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_enabling_statistics.adoc]: again, 'collects' instead of 'exports', here and on next line

File: documentation/src/main/asciidoc/topics/proc_enabling_statistics.adoc [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+<2> exports statistics about the named cache.
+
+[NOTE]
+====
+Enabling statistics for Cache Managers does not globally enable statistics for
+caches. This configuration only exports statistics for the Cache Manager.
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_enabling_statistics.adoc]: collects statistics

File: documentation/src/main/asciidoc/topics/proc_collecting_metrics.adoc [hunk 10]
```diff
@@ -1,45 +1,37 @@
-[id='metric_endpoint']
+[id='collect_metrics']
 = Collecting {brandname} Metrics
-{brandname} servers provide monitoring data through an HTTP `metrics` endpoint
-that exposes OS, JVM, and application-level statistics.
+Collect {brandname} metrics with monitoring tools such as Prometheus.
+
+.Prerequisites
+
+* Enable statistics. If you do not enable statistics, {brandname} provides `0` values for all metrics.
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_collecting_metrics.adoc]: 0 or in some cases -1

File: documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc [hunk 11]
```diff
@@ -0,0 +1,44 @@
+[id='configure_metrics']
+= Enabling {brandname} Metrics
+{brandname} is compatible with the Eclipse MicroProfile Metrics API and can
+generate gauge and histogram metrics.
+
+* {brandname} metrics are provided at the `/base` and `/vendor` scopes and are
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc]: {brandname} metrics are provided at the 'vendor' scope. Also, JVM-related metrics are provided in the 'base' scope.

File: core/src/main/resources/schema/infinispan-config-11.0.xsd [hunk 12]
```diff
@@ -641,21 +645,21 @@
     <xs:attribute name="gauges" type="xs:boolean" default="${GlobalMetrics.gauges}">
       <xs:annotation>
         <xs:documentation>
-          Enables exporting of gauge metrics. These are exported by default if statistics are enabled.
+          Exports gauge metrics. If you enable statistics, gauges are exported by default.
```
Comment (anistor) [inline core/src/main/resources/schema/infinispan-config-11.0.xsd]: I would rephrase ```Exports gauge metrics. Gauges are enabled by default. Statistics must also be enabled or else no metrics will actually be exported. ```

File: core/src/main/resources/schema/infinispan-config-11.0.xsd [hunk 13]
```diff
@@ -641,21 +645,21 @@
...
         </xs:documentation>
       </xs:annotation>
     </xs:attribute>
     <xs:attribute name="histograms" type="xs:boolean" default="${GlobalMetrics.histograms}">
       <xs:annotation>
         <xs:documentation>
-          Enables exporting of histograms. These are costly to compute so are not enabled by default.
+          Exports histograms. Disabled by default because histograms require additional computation.
```
Comment (anistor) [inline core/src/main/resources/schema/infinispan-config-11.0.xsd]: I would rephrase ```Exports histogram metrics. Histograms are not enabled by default because they require additional computation. Statistics must also be enabled or else no metrics will actually be exported. ```

File: core/src/main/resources/schema/infinispan-config-11.0.xsd [hunk 14]
```diff
@@ -672,28 +676,30 @@
...
+            Specifies a JMX property with a name and value that is passed to the MBean Server lookup instance.
+          </xs:documentation>
         </xs:annotation>
       </xs:element>
     </xs:sequence>
     <xs:attribute name="domain" type="xs:string" default="${GlobalJmx.domain}">
       <xs:annotation>
         <xs:documentation>
-          If JMX statistics are enabled then all 'published' JMX objects will appear under this name. This is optional, if not specified an object name will be created for you by default.
+          If JMX statistics are enabled then all 'published' JMX objects  appear under this name. Optional, if not specified a default object name is created.
```
Comment (anistor) [inline core/src/main/resources/schema/infinispan-config-11.0.xsd]: strange space between ```objects``` and ```appear```

File: documentation/src/main/asciidoc/topics/proc_configuring_hotrod_jmx.adoc [hunk 15]
```diff
@@ -13,8 +10,19 @@ Rod client statistics.
 
 [source,java]
 ----
-include::code_examples/HotRodJmx.java[]
+ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
+clientBuilder
+   .statistics()
+    .enable() <1>
+   .jmx()
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_configuring_hotrod_jmx.adoc]: replace ```jmx().enable()``` with ```jmxEnable()```

File: documentation/src/main/asciidoc/topics/proc_configuring_hotrod_jmx.adoc [hunk 16]
```diff
@@ -13,8 +10,19 @@ Rod client statistics.
 
 [source,java]
 ----
-include::code_examples/HotRodJmx.java[]
+ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_configuring_hotrod_jmx.adoc]: An ```import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;``` should make it clear we're not referring to the ConfigurationBuilder from core

File: documentation/src/main/asciidoc/topics/proc_naming_cache_managers.adoc [hunk 17]
```diff
@@ -14,15 +14,21 @@ name, which results in a JMX MBean named `org.infinispan:type=CacheManager,name=
...
 ----
 
 
 .Programmatically
 
 [source,java]
 ----
-include::code_examples/JmxStatisticsCacheManager.java[]
+GlobalConfiguration globalConfig = new GlobalConfigurationBuilder()
+  .jmx().enable()
```
Comment (anistor) [inline documentation/src/main/asciidoc/topics/proc_naming_cache_managers.adoc]: Lines 30 should be before 29 or it won't compile.

**Comment by oraNod** [comment]
> I've added another bunch of comments. I think I'm about 90% done. Please fix these and then I'll read everything again on the rendered doc.

thanks @anistor this is a really good review. I've learned a few things here. the changes are in and I've updated the rendered doc.

**Comment by anistor** [comment]
Merged in master @oraNod ! Thanks!

**Comment by anistor** [comment]
While reviewing the docs (and also looking at what the code actually does, just to make sure) I found this https://issues.redhat.com/browse/ISPN-11463
